types/index.d.ts

```diff
@@ -340,20 +352,59 @@ export interface CookiesOptions {
 }
 
 /** [Documentation](https://next-auth.js.org/configuration/events) */
-export type EventType =
-  | "signIn"
-  | "signOut"
-  | "createUser"
-  | "updateUser"
-  | "linkAccount"
-  | "session"
-  | "error"
+export type EventCallback<MessageType = any> = (
+  message: MessageType
+) => Promise<void>
 
-/** [Documentation](https://next-auth.js.org/configuration/events) */
-export type EventCallback = (message: any) => Promise<void>
+export interface SessionEventMessage {
+  session: Session & Record<string, any>
+  jwt?: JWT
+}
 
-/** [Documentation](https://next-auth.js.org/configuration/events) */
-export type EventsOptions = Partial<Record<EventType, EventCallback>>
+/**
+ * If using a `credentials` type auth, the user is the raw response from your
+ * credential provider.
+ * For other providers, you'll get the User object from your adapter, the account,
+ * and an indicator if the user was new to your Adapter.
+ */
+export interface SignInEventMessage<TUser> {
+  user: TUser
+  account: Record<string, unknown>
+  isNewUser?: boolean
+}
+
+export interface LinkAccountEventMessage<TUser> {
+  user: TUser
+  providerAccount: Record<string, unknown>
+}
+
+export interface EventOptions<TUser> {
```

Comment:
Could we rename this interface to `EventCallbacks`? 

Given all these are callback functions, I think the name is more explicit for developers less familiar with the library 💆🏽‍♂️

I'm aware this will be a breaking change, as before it was also called `EventOptions`, but I think we would gain some clarity renaming it. 

We're treating breaking changes on the types as patch releases of the lib, so I think it would be ok 👍🏽

Reply:
I'm in favor of that rename and agree that breaking the types makes a breaking change overall.